clouddriver-kubernetes-v2/src/main/java/com/netflix/spinnaker/clouddriver/kubernetes/v2/op/handler/KubernetesHorizontalPodAutoscalerHandler.java

```diff
@@ -107,4 +120,105 @@ private static int defaultToZero(@Nullable Integer input) {
+        .forEach(
+            s -> {
+              addResourceKeysForKubernetesResource(mapResourceKeyToManifest, s);
+            });
+
+    for (KubernetesManifest hpa :
+        allResources.getOrDefault(HORIZONTAL_POD_AUTOSCALER, new ArrayList<>())) {
+      relationshipMap.put(
+          hpa,
+          getRelatedManifests(hpa, mapResourceKeyToManifest, deploymentResourceKeysToReplicaSets));
+    }
+  }
+
+  private void addResourceKeysForKubernetesResource(
+      Map<ResourceKey, KubernetesManifest> entries, KubernetesManifest resource) {
+    ResourceKey resourceKey =
+        new ResourceKey(resource.getKind(), resource.getName(), resource.getNamespace());
+    entries.put(resourceKey, resource);
+  }
+
+  private void associateReplicaSetWithDeployments(
+      Map<ResourceKey, List<KubernetesManifest>> deploymentsToReplicaSets,
+      KubernetesManifest replicaSet) {
+    List<ResourceKey> deploymentResourceKeys =
+        replicaSet.getOwnerReferences().stream()
+            .filter(o -> o.getKind().equals(KubernetesKind.DEPLOYMENT))
+            .map(d -> new ResourceKey(d.getKind(), d.getName(), replicaSet.getNamespace()))
+            .collect(Collectors.toList());
+
+    for (ResourceKey deploymentResourceKey : deploymentResourceKeys) {
+      deploymentsToReplicaSets
+          .computeIfAbsent(deploymentResourceKey, d -> new ArrayList<>())
+          .add(replicaSet);
+    }
+  }
+
+  private List<KubernetesManifest> getRelatedManifests(
+      KubernetesManifest hpaManifest,
+      Map<ResourceKey, KubernetesManifest> mapLabelToManifest,
+      Map<ResourceKey, List<KubernetesManifest>> deploymentsToReplicaSets) {
+
+    V1HorizontalPodAutoscaler hpa =
+        KubernetesCacheDataConverter.getResource(hpaManifest, V1HorizontalPodAutoscaler.class);
+
+    String targetRefKind =
+        Optional.ofNullable(hpa.getSpec())
+            .map(V1HorizontalPodAutoscalerSpec::getScaleTargetRef)
+            .map(V1CrossVersionObjectReference::getKind)
+            .orElse("");
+
+    String targetRefName =
+        Optional.ofNullable(hpa.getSpec())
+            .map(V1HorizontalPodAutoscalerSpec::getScaleTargetRef)
+            .map(V1CrossVersionObjectReference::getName)
+            .orElse("");
+
+    String hpaNamespace =
+        Optional.ofNullable(hpa.getMetadata()).map(V1ObjectMeta::getNamespace).orElse("");
+
+    List<KubernetesManifest> result = new ArrayList<>();
+
+    if (targetRefKind.isEmpty() || targetRefName.isEmpty() || hpaNamespace.isEmpty()) {
+      return result;
+    }
+
+    ResourceKey scaleTargetRefResourceKey =
+        new ResourceKey(KubernetesKind.fromString(targetRefKind), targetRefName, hpaNamespace);
+
+    KubernetesManifest matchingReplicaSet = mapLabelToManifest.get(scaleTargetRefResourceKey);
+
+    if (matchingReplicaSet != null) {
+      result.add(matchingReplicaSet);
+    } else {
+      result = deploymentsToReplicaSets.getOrDefault(scaleTargetRefResourceKey, new ArrayList<>());
```

Comment:
I'm still a bit hesitant about doing deployment -> replicaSet resolution here; I remember we talked about this, in particular about how this would affect whether the autoscaler appears on the deployment or on the replicaset, but I'm not sure exactly where we landed.  It feels like if the autoscaler is pointing to a deployment it's really the deployment itself that's being auto-scaled rather than the replicasets.

Aside from the question of whether we should be walking the resource graph to replica sets, if we do I also wonder if it would be clearer to switch on `KubernetesKind` here.  Part of my confusion here is that it's strange that we are trying to look up a replica set, then if we don't find one assuming it's a deployment (which only works because `mapLabelToManifest` does not contain deployments).  It would be clearer if there was a more obvious branching of logic based on the kind.